### PR TITLE
Add "Grass" curve to wolf_sheep example chart

### PR DESCRIPTION
**Pull request by NouamaneTazi**
image

### CODE REVIEW

**Comment by tpike3** [comment]
@NouamaneTazi very cool.  Two requests...

1. Please make sure you are using black and pre-commit (see [contributor guide](https://github.com/projectmesa/mesa/blob/main/CONTRIBUTING.rst) This will stop you builds from failing) This will prevent your builds from failing
2. Could you actually make a separate model example so there is wolf-sheep folder and a wolf-sheep-grass folder. I think this will be beneficial to users to see two interdependent entities  and then three interdependent entities.

File: mesa/time.py [hunk 1]
```diff
@@ -278,8 +278,13 @@ def step_type(self, type_class: Type[Agent], shuffle_agents: bool = True) -> Non
         for agent_key in agent_keys:
             self.agents_by_type[type_class][agent_key].step()
 
-    def get_type_count(self, type_class: Type[Agent]) -> int:
+    def get_type_count(self, type_class: Type[Agent], filter_func: Callable[[Agent], bool] = None) -> int:
```
Comment (rht) [inline mesa/time.py:281]: This is too much of a corner case to warrant putting it in the library? I think it is better for the example to define a new scheduler that inherits from `RandomActivationByType` and overrides this method instead.

File: examples/wolf_sheep/wolf_sheep/model.py [hunk 2]
```diff
@@ -89,6 +89,7 @@ def __init__(
             {
                 "Wolves": lambda m: m.schedule.get_type_count(Wolf),
                 "Sheep": lambda m: m.schedule.get_type_count(Sheep),
+                "Grass": lambda m: m.schedule.get_breed_count(GrassPatch, lambda x: x.fully_grown),
```
Comment (rht) [inline examples/wolf_sheep/wolf_sheep/model.py:92]: Typo: `get_type_count`, not `get_breed_count`.

**Comment by NouamaneTazi** [comment]
> @NouamaneTazi very cool. Two requests...
> 
> 1. Please make sure you are using black and pre-commit (see [contributor guide](https://github.com/projectmesa/mesa/blob/main/CONTRIBUTING.rst?rgh-link-date=2022-03-05T14%3A54%3A46Z) This will stop your builds from failing.)
> 2. Could you actually make a separate model example so there is wolf-sheep folder and a wolf-sheep-grass folder. I think this will be beneficial to users to see two interdependent entities  and then three interdependent entities.

Sorry I didn't pay attention to the pre-commit. Should be fixed by now.
I'm not sure about duplicating the example's folder just for a small chart change. It would probably make more sense to rename the example wolf_sheep_grass, or just leave it as it is (since there is the option to activate/desactivate "Grass" anytime).

**Comment by rht** [comment]
Before it is too late: make sure to add this in your `.git/config`:
```
[pull]
    rebase = true
```

This will prevent the noisy merge commits when you pull from latest main.

**Comment by rht** [comment]
The custom scheduler should be defined inside the examples directory, not core Mesa yet (not until there are demands to put it in core Mesa).

**Comment by rht** [comment]
The examples directory, as in inside the wolf_sheep folder.

**Comment by rht** [comment]
I do think it is totally fine to extend the example to visualize the grass. It's already there.

**Comment by tpike3** [comment]
> I do think it is totally fine to extend the example to visualize the grass. It's already there.

@NouamaneTazi  Sorry about the confusing guidance @rht is correct. I was making an assumption instead of looking at the whole model.

File: examples/wolf_sheep/wolf_sheep/server.py [hunk 3]
```diff
@@ -42,7 +42,11 @@ def wolf_sheep_portrayal(agent):
 
 canvas_element = CanvasGrid(wolf_sheep_portrayal, 20, 20, 500, 500)
 chart_element = ChartModule(
-    [{"Label": "Wolves", "Color": "#AA0000"}, {"Label": "Sheep", "Color": "#666666"}]
+    [
+        {"Label": "Wolves", "Color": "#AA0000"},
+        {"Label": "Sheep", "Color": "#666666"},
+        {"Label": "Grass", "Color": "#00AA00"},
```
Comment (rht) [inline examples/wolf_sheep/wolf_sheep/server.py:48]: I was wondering how the color here is different from https://github.com/projectmesa/mesa/blob/68b6631abd28c2780ab90a3254e6355a3f0de94d/examples/wolf_sheep/wolf_sheep/server.py#L31. Turns out that, that array is a color gradient. See this commit https://github.com/projectmesa/mesa/commit/da194e3556e7a8e61af05da3f726ef42ec7c7c0c.
I think that line is under-documented, and can be clarified by adding a comment that the array is a color gradient.

That said, I am fine with #00AA00.

Comment (rht) [reply examples/wolf_sheep/wolf_sheep/server.py:48]: Raised #1183.

**Comment by rht** [comment]
There is a Readme.md in the `wolf_sheep/` folder. In this readme, there is a list of files and explanation for what each does. And so, the scheduler.py needs to be documented there.

Other than this, LGTM. Make sure to squash the commits into 1 commit.

**Comment by NouamaneTazi** [comment]
> There is a Readme.md in the `wolf_sheep/` folder. In this readme, there is a list of files and explanation for what each does. And so, the scheduler.py needs to be documented there.
> 
> Other than this, LGTM. Make sure to squash the commits into 1 commit.

Fixed. And yes, the commits can be squashed before merging

File: examples/wolf_sheep/Readme.md [hunk 4]
```diff
@@ -7,46 +7,50 @@ A simple ecological model, consisting of three agent types: wolves, sheep, and g
...
-* ``wolf_sheep/schedule.py``: Defines a custom variant on the RandomActivation scheduler, where all agents of one class are activated (in random order) before the next class goes -- e.g. all the wolves go, then all the sheep, then all the grass.
+- ``wolf_sheep/random_walk.py``: This defines the ``RandomWalker`` agent, which implements the behavior of moving randomly across a grid, one cell at a time. Both the Wolf and Sheep agents will inherit from it.
+- ``wolf_sheep/test_random_walk.py``: Defines a simple model and a text-only visualization intended to make sure the RandomWalk class was working as expected. This doesn't actually model anything, but serves as an ad-hoc unit test. To run it, ``cd`` into the ``wolf_sheep`` directory and run ``python test_random_walk.py``. You'll see a series of ASCII grids, one per model step, with each cell showing a count of the number of agents in it.
+- ``wolf_sheep/agents.py``: Defines the Wolf, Sheep, and GrassPatch agent classes.
+- ``wolf_sheep/scheduler.py``: Defines a custom variant on the RandomActivationByType scheduler, where we can define filters for
+
+- the `get_type_count` function.
```
Comment (rht) [inline examples/wolf_sheep/Readme.md:43]: This is wrongly formatted.

File: examples/wolf_sheep/Readme.md [hunk 5]
```diff
@@ -7,46 +7,50 @@ A simple ecological model, consisting of three agent types: wolves, sheep, and g
 If wolves and sheep have enough energy, they reproduce, creating a new wolf or sheep (in this simplified model, only one parent is needed for reproduction). The grass on each cell regrows at a constant rate. If any wolves and sheep run out of energy, they die.
 
 The model is tests and demonstrates several Mesa concepts and features:
- - MultiGrid
- - Multiple agent types (wolves, sheep, grass)
- - Overlay arbitrary text (wolf's energy) on agent's shapes while drawing on CanvasGrid
- - Agents inheriting a behavior (random movement) from an abstract parent
- - Writing a model composed of multiple files.
- - Dynamically adding and removing agents from the schedule
+
```
Comment (rht) [inline examples/wolf_sheep/Readme.md:10]: This empty line is not necessary.

**Comment by rht** [comment]
LGTM. The squash and merge button is disabled, so it would be great if you could squash the commits into 1.

**Comment by rht** [comment]
Merged, thank you @NouamaneTazi !

**Pull request merged**